# Movies Tonight Part 3: SQL DML

In [9]:
%load_ext sql
%sql sqlite:///Movies.db

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


'Connected: @Movies.db'

In [10]:
%%sql
SELECT * FROM DATASET LIMIT 10;
PRAGMA table_info(MOVIES);

 * sqlite:///Movies.db
Done.
Done.


[(0, 'MID', 'INTEGER', 0, None, 1),
 (1, 'TITLE', 'TEXT', 1, None, 0),
 (2, 'RATING', 'TEXT', 0, None, 0)]

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

In [32]:
%%sql
DROP TABLE IF EXISTS THEATRE;

DELETE FROM MOVIES;
INSERT INTO MOVIES (TITLE, RATING)
    SELECT DISTINCT MTitle, Rating FROM DATASET;
SELECT * FROM MOVIES;

DELETE FROM THEATERS;
INSERT INTO THEATERS (Name, Location, Phone)
    SELECT DISTINCT TName, Location, Phone FROM DATASET;
SELECT * FROM THEATERS;

DELETE FROM ARTISTS;
INSERT INTO ARTISTS (NAME)
    SELECT DISTINCT CName FROM DATASET;
SELECT * FROM ARTISTS
LIMIT 5;

DELETE FROM SHOWS;
INSERT INTO SHOWS (SHOWTIME, MID, TID)
    SELECT DISTINCT DATASET.ShowTime, MOVIES.MID, THEATERS.TID
    FROM DATASET
        JOIN MOVIES ON (MOVIES.TITLE = DATASET.MTitle)
        JOIN THEATERS ON (THEATERS.NAME = DATASET.TName);
SELECT SHOWTIME, TITLE, NAME FROM SHOWS JOIN MOVIES USING (MID) JOIN THEATERS USING (TID);

DELETE FROM CREDIT;
INSERT INTO CREDIT (CCode, AID, MID)
    SELECT DISTINCT CCode, AID, MID
    FROM DATASET
        JOIN ARTISTS ON (ARTISTS.NAME = DATASET.CName)
        JOIN MOVIES ON (MOVIES.TITLE = DATASET.MTitle);
SELECT * FROM CREDIT;


 * sqlite:///Movies.db
Done.
23 rows affected.
23 rows affected.
Done.
9 rows affected.
9 rows affected.
Done.
153 rows affected.
153 rows affected.
Done.
131 rows affected.
131 rows affected.
Done.
0 rows affected.
162 rows affected.
Done.


[(1, 1, 'A', 1),
 (2, 2, 'A', 1),
 (3, 3, 'A', 1),
 (4, 4, 'A', 1),
 (5, 5, 'A', 1),
 (6, 6, 'A', 1),
 (7, 7, 'A', 1),
 (8, 8, 'A', 1),
 (9, 9, 'D', 1),
 (10, 10, 'A', 2),
 (11, 11, 'A', 2),
 (12, 12, 'A', 2),
 (13, 13, 'A', 2),
 (14, 14, 'A', 2),
 (15, 15, 'A', 2),
 (16, 16, 'A', 2),
 (17, 17, 'A', 2),
 (18, 18, 'D', 2),
 (19, 19, 'A', 3),
 (20, 20, 'A', 3),
 (21, 21, 'A', 3),
 (22, 22, 'A', 3),
 (23, 23, 'A', 3),
 (24, 24, 'A', 3),
 (25, 25, 'A', 3),
 (26, 26, 'A', 3),
 (27, 27, 'A', 3),
 (28, 28, 'A', 3),
 (29, 29, 'A', 3),
 (30, 30, 'D', 3),
 (31, 31, 'A', 4),
 (32, 32, 'A', 4),
 (33, 33, 'A', 4),
 (34, 34, 'A', 4),
 (35, 35, 'A', 4),
 (36, 35, 'D', 4),
 (37, 36, 'A', 5),
 (38, 37, 'A', 5),
 (39, 38, 'A', 5),
 (40, 39, 'A', 5),
 (41, 40, 'A', 5),
 (42, 41, 'A', 5),
 (43, 42, 'A', 5),
 (44, 43, 'D', 5),
 (45, 44, 'A', 6),
 (46, 45, 'A', 6),
 (47, 46, 'A', 6),
 (48, 47, 'D', 6),
 (49, 48, 'A', 7),
 (50, 49, 'A', 7),
 (51, 50, 'A', 7),
 (52, 51, 'A', 7),
 (53, 52, 'D', 7),
 (54, 53, 'A', 8),
 (55, 54, 'A', 8),
 (56, 55, 'A', 8),
 (57, 56, 'A', 8),
 (58, 57, 'A', 8),
 (59, 58, 'A', 8),
 (60, 59, 'D', 8),
 (61, 60, 'A', 9),
 (62, 61, 'A', 9),
 (63, 62, 'A', 9),
 (64, 63, 'A', 9),
 (65, 60, 'D', 9),
 (66, 64, 'A', 10),
 (67, 65, 'A', 10),
 (68, 66, 'A', 10),
 (69, 67, 'A', 10),
 (70, 68, 'A', 10),
 (71, 69, 'D', 10),
 (72, 70, 'A', 11),
 (73, 71, 'A', 11),
 (74, 72, 'A', 11),
 (75, 73, 'A', 11),
 (76, 74, 'A', 11),
 (77, 75, 'A', 11),
 (78, 76, 'A', 11),
 (79, 77, 'A', 11),
 (80, 78, 'A', 11),
 (81, 79, 'A', 11),
 (82, 80, 'A', 11),
 (83, 81, 'A', 11),
 (84, 82, 'D', 11),
 (85, 53, 'A', 12),
 (86, 83, 'A', 12),
 (87, 84, 'A', 12),
 (88, 85, 'A', 12),
 (89, 86, 'D', 12),
 (90, 87, 'A', 13),
 (91, 88, 'A', 13),
 (92, 89, 'A', 13),
 (93, 90, 'A', 13),
 (94, 91, 'A', 13),
 (95, 92, 'D', 13),
 (96, 93, 'A', 14),
 (97, 94, 'A', 14),
 (98, 95, 'A', 14),
 (99, 96, 'A', 14),
 (100, 97, 'A', 14),
 (101, 98, 'D', 14),
 (102, 99, 'A', 15),
 (103, 100, 'A', 15),
 (104, 101, 'A', 15),
 (105, 102, 'A', 15),
 (106, 103, 'A', 15),
 (107, 104, 'A', 15),
 (108, 105, 'D', 15),
 (109, 106, 'A', 16),
 (110, 107, 'A', 16),
 (111, 108, 'A', 16),
 (112, 77, 'A', 16),
 (113, 109, 'A', 16),
 (114, 110, 'A', 16),
 (115, 106, 'D', 16),
 (116, 109, 'D', 16),
 (117, 111, 'A', 17),
 (118, 112, 'A', 17),
 (119, 113, 'A', 17),
 (120, 114, 'D', 17),
 (121, 115, 'A', 18),
 (122, 116, 'A', 18),
 (123, 117, 'A', 18),
 (124, 118, 'A', 18),
 (125, 119, 'A', 18),
 (126, 120, 'A', 18),
 (127, 121, 'D', 18),
 (128, 122, 'A', 19),
 (129, 123, 'A', 19),
 (130, 124, 'A', 19),
 (131, 125, 'A', 19),
 (132, 126, 'D', 19),
 (133, 127, 'A', 20),
 (134, 117, 'A', 20),
 (135, 128, 'A', 20),
 (136, 129, 'A', 20),
 (137, 130, 'A', 20),
 (138, 28, 'A', 20),
 (139, 131, 'D', 20),
 (140, 132, 'A', 21),
 (141, 100, 'A', 21),
 (142, 133, 'A', 21),
 (143, 134, 'A', 21),
 (144, 135, 'A', 21),
 (145, 136, 'D', 21),
 (146, 137, 'A', 22),
 (147, 138, 'A', 22),
 (148, 139, 'A', 22),
 (149, 140, 'A', 22),
 (150, 141, 'A', 22),
 (151, 142, 'A', 22),
 (152, 143, 'A', 22),
 (153, 144, 'A', 22),
 (154, 145, 'A', 22),
 (155, 146, 'A', 22),
 (156, 147, 'D', 22),
 (157, 148, 'A', 4),
 (158, 149, 'A', 23),
 (159, 150, 'A', 23),
 (160, 151, 'A', 23),
 (161, 152, 'A', 23),
 (162, 153, 'D', 23)]

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [ ]:
YOUR CODE HERE

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

In [ ]:
YOUR CODE HERE

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [ ]:
YOUR CODE HERE

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".